### PR TITLE
Cache Yarn Dependencies in GitHub Actions

### PR DESCRIPTION
**Pull request by noi5e**
See #9983 particularly [this comment](https://github.com/publiclab/plots2/issues/9983#issuecomment-894716371).

Work in progress, first I'm going to try to echo the yarn cache directory.

### CODE REVIEW

**Comment by noi5e** [comment]
Okay, that made this happen:
Screen Shot 2021-08-07 at 4 21 16 PM

I was expecting the directory to be printed in the run.

Not really sure what I'm doing, maybe I should try failing harder. I'll try implementing this part from [the docs](https://github.com/actions/cache/blob/main/examples.md#node---yarn):

```
- uses: actions/cache@v2
  id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
  with:
    path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
    key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
    restore-keys: |
      ${{ runner.os }}-yarn-
```

**Comment by noi5e** [comment]
Functional tests failed after my last changes:
Screen Shot 2021-08-07 at 4 34 47 PM

Here's the error:
Screen Shot 2021-08-07 at 4 34 12 PM

`Cache not found for input keys`. 🤔 

- There's a bit `key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}` that I added. Maybe this needs to be a more specific key?
- I also removed the `yarn install` part, maybe I'm not supposed to do that?

**Comment by noi5e** [comment]
@jywarren Whenever it makes sense, can you take a look at this? 😄

**Comment by jywarren** [comment]
Thanks so much for looking into this! As you can tell i'm a little behind on reviews, will catch up here ASAP!

**Comment by noi5e** [comment]
Latest commit: trying to run `yarn cache dir` according to [these docs](https://classic.yarnpkg.com/en/docs/cli/cache/#toc-yarn-cache-dir) instead of the other command.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -57,6 +57,16 @@ jobs:
     - uses: ./.github/actions/setup-test-environment
     - name: Install JavaScript dependencies with Yarn
       run: yarn check || yarn install --frozen-lockfile;
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      run: yarn cache dir
+    - uses: actions/cache@v2
+      id: yarn-cache # use this to check for `cache-hit` (`steps.yarn-cache.outputs.cache-hit != 'true'`)
+      with:
+        path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
```
Comment (jywarren) [inline .github/workflows/tests.yml:66]: Huh, seeing an error here:
 
```
Run actions/cache@v2
  with:
    key: Linux-yarn-08fd7bab35005d3e8e92ed89e54f7a999ee6535e35e43d54deb4a9a5efd3bf37
    restore-keys: Linux-yarn-
  
Error: Input required and not supplied: path
```

Comment (jywarren) [reply .github/workflows/tests.yml:66]: I'm a little stumped, to be honest, sorry! 

But, your `yarn cache dir` seems to have output something useful:

```
Run yarn cache dir
  yarn cache dir
  shell: /usr/bin/bash -e {0}
/home/runner/.cache/yarn/v6
```

Comment (jywarren) [reply .github/workflows/tests.yml:66]: Ah so do we need to define an output of name `dir` for `yarn-cache-dir-path`?

Comment (jywarren) [reply .github/workflows/tests.yml:66]: aha -- https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#outputs-for-composite-actions

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -57,6 +57,20 @@ jobs:
     - uses: ./.github/actions/setup-test-environment
     - name: Install JavaScript dependencies with Yarn
       run: yarn check || yarn install --frozen-lockfile;
+    - name: Get yarn cache directory path
+      id: yarn-cache-dir-path
+      outputs:
```
Comment (jywarren) [inline .github/workflows/tests.yml:62]: Tried this!

Comment (jywarren) [reply .github/workflows/tests.yml:62]: Huh... an hour later tests haven't started...?

**Comment by jywarren** [comment]
Is something here blocking GitHub action runs? Hmm...

**Comment by jywarren** [comment]
Aha - under Checks, i see this error:

```
Invalid workflow file : .github/workflows/tests.yml#L62
The workflow is not valid. .github/workflows/tests.yml (Line: 62, Col: 7): Unexpected value 'outputs'
```

**Comment by jywarren** [comment]
Could we just use this? https://github.com/marketplace/actions/yarn-node-install-cache

**Comment by jywarren** [comment]
Well, that ran! I see:

```

/usr/bin/uname
Linux
/usr/local/bin/yarn cache dir
/home/runner/.cache/yarn/v6
Cache keys: Linux-yarn-node-cache-2fee04878e10de47453d7a3168291c15, Linux-yarn-node-cache-
Cache not found
```

but, the first time, the cache would be generated, right? I'm not sure the cache persistence rules. I'll try running this again to see if it finds the cache from this last run?

**Comment by jywarren** [comment]
Unfortunately it doesn't save the previous run if we don't make a new commit (or maybe i just don't know how): https://github.com/publiclab/plots2/actions/runs/1195643414

**Comment by jywarren** [comment]
Wait! Yay! It worked!!!

```
Run egordm/gha-yarn-node-cache@v1
/usr/bin/uname
Linux
/usr/local/bin/yarn cache dir
/home/runner/.cache/yarn/v6
Cache keys: Linux-yarn-node-cache-2fee04878e10de47453d7a3168291c15, Linux-yarn-node-cache-
Cache Size: ~242 MB (253387137 B)
/usr/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/9a3235da-b161-4a2e-a4fd-a1baac79bd2e/cache.tzst -P -C /home/runner/work/plots2/plots2
Cache restored from key: Linux-yarn-node-cache-2fee04878e10de47453d7a3168291c15
```

Wow i almost didn't expect that to work. 

Does this mean we should be OK now? 

Hmm. When i look at the next step, "Install JavaScript dependencies with Yarn" i see that the latest run took only 21s vs. the first commit in this PR taking 46s. So that seems good although maybe not as good as I'd have guessed. 

But, when I look closer, I see that both end with similar times:

* the first commit cites 14.94s - https://github.com/publiclab/plots2/runs/3271177757#step:7:2243
* the latest cites 16.20s to completion - https://github.com/publiclab/plots2/pull/9998/checks?check_run_id=3500056847#step:7:91

However, the first commit has about 2000 lines of errors in the `yarn check` output, while the latest has no errors. So `yarn check` accounts for the discrepancy actually, not `yarn install --frozen-lockfile`!

The command for install is, for reference: `yarn check || yarn install --frozen-lockfile;`

**Comment by jywarren** [comment]
Overall here are the times: 

First commit: 

image

Latest commit:

image

I'm re-running now for comparison.

**Comment by jywarren** [comment]
Hmm. I mean, ~22 seconds for yarn install consistently. But, that's not really making a dent in our overall run time: 

image

**Comment by jywarren** [comment]
I mean, i guess that step is just not that lengthy anyways. But nice to have the cache running if I read this right. @noi5e what do you think?

**Comment by jywarren** [comment]
I guess let's copy this into the other steps too. 

Overall our startup and setup aren't too bad: 

image

because we're parallelized, shaving 20s didn't have a huge effect.

**Pull request merged**